### ISSUE TITLE
Request for Beam Search Understanding in FT Code

### ISSUE THREAD

**Issue by HossamAmer12** [NVIDIA/FasterTransformer#346]
Hi, 

What I know about the beam search step is that it does the following code in python

 `def step(step: int, lprobs, scores):
    bsz, beam_size, vocab_size = lprobs.size()
    if step == 0:
        lprobs = lprobs[:, ::beam_size, :].contiguous()
    else:
        # make probs contain cumulative scores for each hypothesis
        lprobs = lprobs + scores[:, :, step - 1].unsqueeze(-1)
    top_prediction = torch.topk(
        lprobs.view(bsz, -1),
        k=min(
            # Take the best 2 x beam_size predictions. We'll choose the first
            # beam_size of these which don't predict eos to continue with.
            beam_size * 2,
            lprobs.view(bsz, -1).size(1) - 1,  # -1 so we never select pad
        ),
    )
    scores_buf = top_prediction[0]
    indices_buf = top_prediction[1]
    # Project back into relative indices and beams
    beams_buf = indices_buf // vocab_size
    indices_buf = indices_buf.fmod(vocab_size)
    # At this point, beams_buf and indices_buf are single-dim and contain relative indices
    return scores_buf, indices_buf, beams_buf` 
    
    In FT:
    1- beam_online_softmax_topk_stage1_kernel: I see you add the bias to the logits. Why is reduce_md_op useful in this case? I don't understand the calculation. Why do you subtract the bigger from the smaller and multiply by the exponent? How's this relevant to logsoftmax?
    2- beam_online_softmax_topk_stage1_kernel: float d_total_log = logf(total.md.d); - What's this doing?
    3- beam_online_softmax_topk_stage1_kernel: float val = total.topk.u[i] - total.md.m - d_total_log; - What's this doing?
    4- beam_online_softmax_topk_stage2_kernel: What is the role of this kernel?
    5- batch_topK_kernel: Why do you call this after stage2?
   
    I would really appreciate your help in understanding this. The code is very efficient, but hard to understand relative to the python code above.

**Comment by byshiue** [comment]
The implementation refers online softmax in https://github.com/NVIDIA/online-softmax

**Comment by HossamAmer12** [comment]
Thanks for sharing the paper @byshiue .. I will take a look :)

**Comment by HossamAmer12** [comment]
@byshiue In Fairseq, there are methods for active hypos and finalize hypos, which are illustrated here:
https://github.com/facebookresearch/fairseq/blob/1f533150710a3b393650dc4e3998c3675e42e3c9/fairseq/sequence_generator.py#L565

Do you do the same in your beam search? So far, I cannot see that you do that.
